src/utils/rari.js

```diff
@@ -0,0 +1,237 @@
+  return mstableContract.getSwapOutput(
+    inputTokenAddress,
+    outputTokenAddress,
+    amountBN,
+  ).catch((error) => {
+    reportErrorLog("Rari service failed: Can't get mStable swap output", { error });
+    return null;
+  });
+};
+
+const getRariDepositTransactionData = async (
+  amountBN: EthersBigNumber, token: Asset, supportedAssets: Asset[], rates: Rates,
+) => {
+  const acceptedCurrencies = await getRariAcceptedCurrencies();
+  if (!acceptedCurrencies) return null;
+
+  let txValue = EthersBigNumber.from(0);
+  if (token.symbol === ETH) {
+    txValue = txValue.add(amountBN);
+  }
+
+  // token can be directly deposited, no exchange fee
+  if (acceptedCurrencies.includes(token.symbol)) {
+    return {
+      depositTransactionData: encodeContractMethod(RARI_FUND_MANAGER_CONTRACT_ABI, 'deposit', [
+        token.symbol,
+        amountBN,
+      ]),
+      txValue,
+      exchangeFeeBN: EthersBigNumber.from(0),
+      slippage: 0,
+      rariContractAddress: getEnv().RARI_FUND_MANAGER_CONTRACT_ADDRESS,
+    };
+  }
+
+  let acceptedCurrency = acceptedCurrencies[0];
+  let acceptedAsset = supportedAssets.find(asset => asset.symbol === acceptedCurrency);
+
+  if (!acceptedAsset) {
+    reportErrorLog("Rari service failed: can't find asset", { asset: acceptedCurrency });
+    return null;
+  }
+
+  // try mStable
+  if ([DAI, USDC, USDT, TUSD].includes(token.symbol)) {
+    for (let i = 0; i < acceptedCurrencies.length; ++i) {
+      acceptedCurrency = acceptedCurrencies[i];
+      acceptedAsset = supportedAssets.find(asset => asset.symbol === acceptedCurrencies[i]);
+
+      if (![DAI, USDC, USDT, TUSD, mUSD].includes(acceptedCurrency) || !acceptedAsset) {
+        continue;  // eslint-disable-line
+      }
+
+      // eslint-disable-next-line
+      const mStableSwapOutput = await getMStableSwapOutput(token.address, acceptedAsset.address, amountBN);
+      if (mStableSwapOutput && mStableSwapOutput['0']) {
+        const ethRates = rates[ETH];
+        if (!ethRates) {
+          reportErrorLog('Rari service failed: no ETH exchange rate');
+        }
+        const formattedInput = utils.formatUnits(amountBN, token.decimals);
+        const formattedOutput = utils.formatUnits(mStableSwapOutput.output, acceptedAsset.decimals);
+        const mStableFeeUSD = formattedInput - formattedOutput; // we assume that stablecoins price is 1$
+
+        return {
+          depositTransactionData: encodeContractMethod(
+            RARI_FUND_PROXY_CONTRACT_ABI,
+            'exchangeAndDeposit(string,uint256,string)',
+            [
+              token.symbol,
+              amountBN,
+              acceptedCurrency,
+            ]),
+          txValue,
+          exchangeFeeBN: utils.parseEther((mStableFeeUSD / ethRates[USD]).toFixed(18)),
+          slippage: 0, // mStable has no slippage
+          rariContractAddress: getEnv().RARI_FUND_PROXY_CONTRACT_ADDRESS,
+        };
+      }
+    }
+  } else if (token.symbol === mUSD) {
```

Comment:
What's up here?

Reply:
The comment should explain everything... The flow is a little bit different for mUSD, but since mUSD is not supported by Pillar wallet, there is no way for the user to select it. So for now it makes no sense to implement it, but I added a comment just in case it will be added in the future.

Reply:
So I think we can add this at the top of logic as "TODO" and remove empty "if" statement as it looks that something is missing even though comment is inside, wdyt?